Implement core Fog Overseer functionality

samdealy opened the pull request:
### Motivation
This PR contains the core functionality of the Fog Overseer service. The service monitors the active Fog Ingest cluster and makes sure that there is at least one active node. If there is no active node, then it performs the automatic failover logic. This logic is described in detail in the worker.rs documentation. 

Note that this PR contains the core logic: the monitoring logic is in worker.rs and the rest of the files basically just expose that functionality to clients via a Rocket HTTPS server. The tests test 5 different scenarios in an end to end fashion. Note that there are still some key features that need to be implemented before we can deploy this service. Namely, we'll need to implement the following items: 

1. Sentry alerting: in some cases, we need to send alerts to humans when Overseer encounters certain (unlikely) Fog Ingest cluster states.
2. Prometheus metrics: as noted in the design doc, we want to capture Fog ingest cluster state metrics like number of ingress keys, number of responsive Fog ingest nodes, etc. 
3. Get info api: in addition to storing Fog ingest cluster state in Prometheus, we want to expose a get info api that allows clients to get the current Fog Ingest cluster info in real time.
4. Config refresh: currently, we pass the list of Fog Ingest node uris to overseer through a CLI config. To allow for overseer to look at a new set of node urls during blue / green deployment without having to restart it, we'll move this list of uris to an environment variable that can be dynamically refreshed. This way, we can update the list of uris without having to restart overseer.

I figured that it would be best if I implemented these 4 features in future PRs, since this PR is already quite large. However, if you find that any of these features should be in this PR, I'm open to discussing. 

### In this PR
* Implements the core functionality of the Fog Overseer Service
* Creates 5 end to end tests that cover the core scenarios we expect Fog Overseer to help with.

### Future Work
* Implement the 4 features listed above.

wjuan-mob commented on fog/overseer/server/src/service.rs:
```diff
@@ -0,0 +1,80 @@
+        ret
+    }
+
+    fn start_helper(&mut self) -> Result<(), OverseerError> {
```
Why split this out from start?

samdealy replied:
It helps keep the start function small and focused- it deals with handling any errors that might arise from starting the service without getting into the nitty gritty details of what threads etc. are started. Moreover, splitting the logic out between a `start` and `start_helper` is the convention for Fog services.

wjuan-mob commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,546 @@
+where
+    OverseerError: From<DB::Error>,
+{
+    /// Poll the Fog Ingest cluter every 5 seconds.
```
Nit: Typo

samdealy replied:
Done.

wjuan-mob commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,546 @@
+
+    fn run(self) {
+        while !self.stop_requested.load(Ordering::SeqCst) {
+            log::trace!(self.logger, "Overser worker start of thread.");
```
Nit: typo

samdealy replied:
Done.

wjuan-mob commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,546 @@
+            should_include_lost_keys: false,
+            // A retired key can be outstanding if public expiry is greater than
+            // its last scanned block.
+            should_include_retired_keys: true,
```
The comment on should include retired keys and should_only_include_unexpired_keys is a bit confusing. Could we clarify this a little bit? How are these 2 different?

samdealy replied:
The two filters are different in that the `should_include_retired_keys` set to `true` includes keys that have been marked retired in the recovery DB (setting it to `false` discards any key that is marked retired). The second filter, `should_only_include_unexpired_keys` discards any key that is expired- an expired key is one in which it's pubkey_expiry (i.e. the last block for which the key must be used to scan) is <= the last scanned block. 

 We need retired keys because they can still be outstanding, meaning they might still have work to do because their pubkey_expiry (i.e. the last block for which they are responsible for scanning) might be greater than the last scanned block.

 I added to the `should_include_retired_keys` comment to try and clarify this point. Let me know if that helps or if you'd like to see anything else written in this comment.

wjuan-mob commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,546 @@
+    }
+
+    fn set_new_key_on_a_node(&self) -> Result<usize, OverseerError> {
+        for (i, ingest_client) in self.ingest_clients.iter().enumerate() {
```
I think we could assert that we're only calling this on idle nodes as a precaution.

samdealy replied:
I need to think about putting in an assert for idle state. I agree that it might be good for precaution's sake, but this method is only called when the main method, `perform_automatic_failover`, has determined that every node for which it received a response, is idle. 

I guess the question is: Is it possible for one of these idle nodes to become active in between when `perform_automatic_failover` determines that it's idle and when this method is called? In a perfect world, the answer is no. I'll think more on it to see if there any cases that might violate this assumption. If there are, then the assert makes perfect sense. 

@garbageslam any thoughts on this?

garbageslam replied:
I think the server already checks this basically -- it won't allow you to call new_keys if the ingest server is in the active state. I think that's the right layer to guard against this.

That being said, it probably can't hurt to put an assert here though. But I don't think it's essential.

samdealy replied:
Yeah, agreed that the server layer is the right layer to check this. I'll leave as is for now, but if we decide to change in the future, will do.

wjuan-mob commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,546 @@
+    }
+
+    fn activate_a_node(&self, activated_node_index: usize) -> Result<(), OverseerError> {
+        let result = retry_with_index(Fixed::from_millis(200), |current_try| {
```
This should probably also only be called on idle nodes.

samdealy replied:
Same as above.

wjuan-mob reviewed (approved):
I think it might be nice to move more of the boilerplate in the tests into the test utils or have a macro. But I don't think it's necessary. 

Is it possible for a concurrency issue to occur where the overseer queries the servers in sequence for their state and one becomes active after being queried but before the others are done. So then the overseer believes that they are all idle and attempts to do something.

eranrund reviewed (changes requested):
Nice work.
I have not looked at the tests, but I did have some (relatively minor) suggestions, and an open question about the use of stop_trigger.
I think what should've happened is that arming starts the worker thread or errors if it is already started, and disarming stops it?

samdealy commented:
> I think it might be nice to move more of the boilerplate in the tests into the test utils or have a macro. But I don't think it's necessary.
I think this is a good point, and we should start having conversations as a team to see what we should and shouldn't include in tests / test utils. I tend to follow the convention that tests should be DAMP rather than DRY (https://testing.googleblog.com/2019/12/testing-on-toilet-tests-too-dry-make.html). I try to include enough stuff in a test such that the reader doesn't need to go to another file to understand what is going on in the test. For some things, like setting up the Ingest node, I make a method for it because the reader doesn't need to know how the node is made. 

> Is it possible for a concurrency issue to occur where the overseer queries the servers in sequence for their state and one becomes active after being queried but before the others are done. So then the overseer believes that they are all idle and attempts to do something.

I believe it's pretty unlikely that this would happen because the overseer is started after the three nodes are started, and the first node is activated. However, if we wanted to be very certain about this sequencing, we could implement some sort of async / await or condition variable that allows the test to wait until the nodes are started and activated before starting overseer.

wjuan-mob commented:
> > I think it might be nice to move more of the boilerplate in the tests into the test utils or have a macro. But I don't think it's necessary.
> 
> I think this is a good point, and we should start having conversations as a team to see what we should and shouldn't include in tests / test utils. I tend to follow the convention that tests should be DAMP rather than DRY (https://testing.googleblog.com/2019/12/testing-on-toilet-tests-too-dry-make.html). I try to include enough stuff in a test such that the reader doesn't need to go to another file to understand what is going on in the test. For some things, like setting up the Ingest node, I make a method for it because the reader doesn't need to know how the node is made.
> 
> > Is it possible for a concurrency issue to occur where the overseer queries the servers in sequence for their state and one becomes active after being queried but before the others are done. So then the overseer believes that they are all idle and attempts to do something.
> 
> I believe it's pretty unlikely that this would happen because the overseer is started after the three nodes are started, and the first node is activated. However, if we wanted to be very certain about this sequencing, we could implement some sort of async / await or condition variable that allows the test to wait until the nodes are started and activated before starting overseer.

I generally agree with damp over dry. I just think that some amount of the boilerplate in these tests is irrelevant to the actual condition that is being tested, and extracting it might make it easier to read. Basically everything prior to when nodes start being set for example.

samdealy commented on fog/overseer/server/Cargo.toml:
```diff
@@ -0,0 +1,50 @@
+url = "2.2.2"
+
+# mc
+mc-attest-net = { path = "../../../attest/net" }
+mc-watcher = { path = "../../../watcher" }
+mc-ledger-db = { path = "../../../ledger/db" }
+mc-util-from-random = { path = "../../../util/from-random" }
```
Done.

samdealy commented on fog/overseer/server/src/bin/main.rs:
```diff
@@ -0,0 +1,42 @@
+    let config = OverseerConfig::from_args();
+    let (logger, _global_logger_guard) = mc_common::logger::create_app_logger(
+        o!("mc.local_overseer_node_id" => config.local_overseer_node_id.to_string()),
+    );
```
Added a comment.

samdealy commented on fog/overseer/server/src/config.rs:
```diff
@@ -0,0 +1,58 @@
+    #[structopt(long)]
+    pub local_overseer_node_id: ResponderId,
+
+    /// TODO: Make this an environment variable that can be dynamically
```
For context, this idea came up in a discussion with ops about how to keep one Fog Overseer instance running between blue-green deployments. 

I had, perhaps naively, thought that there'd be a widely-used, OOTB way to achieve this in Rust. After digging in a little more, I found that this isn't the case, but there is this crate, [spirit](https://docs.rs/spirit/0.3.7/spirit/), that appears to allow for config reloading. What do you think about this crate? Not sure if we've looked into this before? 

@jgreat Feel free to add anything re: our discussion on dynamic config reloading.

eranrund replied:
Why not send configuration changes over an HTTP API?

samdealy replied:
I tend to think that configuration shouldn't be modified over HTTP and that we should try to keep configs in a text file. That way, anyone can look at that text file at any given time to see what the configs are. However, I'd be happy to do HTTP given it's simplicity / if ops is ok with it.

eranrund replied:
Text file is different than an env variable though. I view env variable as "dynamic" in the same way I'd consider HTTP dynamic, and a text file as "static"

wjuan-mob replied:
Could you not save the active configs to a text file even if it's loaded over an http api?

samdealy replied:
Ah, I was using text file == env variable because most env variables are set in some sort of .env file. The way I envision it, updates to configs would be made by editing the text file (making it dynamic and not static). Then a worker would check for an updated config. 

You could also save the active configs to a text file, but at that point why not just have the person changing the config edit the text file directly?

samdealy replied:
I'm curious as to what @jgreat thinks about this. Would you prefer to update the list of Fog Ingest nodes that Overseer watches through (a) HTTP request or (b) updating a config file and having Overseer check for these updates?

garbageslam replied:
Maybe it could do something with like, a shared volume i guess? That might be durable? Lets see what @jgreat says

samdealy replied:
@joekottke this is the comment I mentioned offline.

joekottke replied:
@garbageslam @samdealy We can very easily map in a config into the running environment.  It will also follow a given running instance around even if k8s moves the running instance to another machine.  The file effectively gets mapped in as an external volume.  Another nice effect of this is that we can _change_ the running config file, k8s will update that file on the running instance(s) (after a few seconds as changes get pushed out).  The on-disk semantics of this are handled through symlinks:  The running config is mapped via a symlink to a file, and then the new file gets mapped in, and the symlink gets updated.  One caveat: if the 'config file watcher' is only watching a specifc on-disk inode (after following the symlink reference) then it _might not_ notice the updated file (because it is at a new inode).  Just something to keep in mind.

joekottke replied:
Some other things to keep in mind:
* k8s will have to reference the same config file contents for both the active and inactive nodes, at least the way we have things deployed now.
* The config file is read-only

samdealy commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,546 @@
+                        Err(_) => {
+                            // TODO: See if we can make an id field for ingest client. This
+                            // will allow us to better identify ingest clients.
+                            log::warn!(self.logger, "Could not activate idle ingest client.");
```
IIUC, yes, I opted to treat the node's URI as it's identifier. If using the URI as an ID is fine, I can just remove this comment.

eranrund replied:
I think the URI is a good unique identifier for ingest nodes.

samdealy replied:
SG, removed the comment.

samdealy commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,546 @@
+            }
+        }
+
+        Err(OverseerError::ActivateNode("Could not activate a node because no node has an ingress key that matches the outstanding key.".to_string()))
```
Done.

samdealy commented on fog/overseer/server/src/config.rs:
```diff
@@ -0,0 +1,58 @@
+
+    /// Local Ingest Node ID
+    #[structopt(long)]
+    pub local_overseer_node_id: ResponderId,
```
Ah, I didn't realize that ResponderId corresponded to an enclave. I still want to have some sort of ID that corresponds to the Fog Overseer instance (one day, we'll be able to run multiple overseer instances at once), so perhaps a simple String would suffice instead of a ResponderId?

eranrund replied:
Is the hostname of the machine not good enough?

eranrund replied:
I'd also suggest that there's no need to future-proof this at the moment. Multiple instances will be a big enough change that this can be introduced at that point.

samdealy replied:
SG, just removed the usage of ResponderId.

samdealy commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,546 @@
+        inactive_outstanding_key: CompressedRistrettoPublic,
+    ) -> Result<(), OverseerError> {
+        let result = retry_with_index(Fixed::from_millis(200), |current_try| {
+            if current_try as u8 > Self::NUMBER_OF_TRIES {
```
I used it because I'd like the error message to include how many tries remain. I don't think the take option enables this? If it does, then great- it's more compact and would change to it.

Relevant [documentation](https://docs.rs/retry/latest/retry/#usage):
`If your operation needs to know how many times it’s been tried, use the retry_with_index function. This works the same as retry, but passes the number of the current try to the closure as an argument.`

eranrund replied:
You can use a combination of `retry_with_index` and `take`, which will allow you to include the number of retries remaining. It will save you from having to implement the limiting yourself.

samdealy commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,546 @@
+            }
+        });
+
+        match result {
```
Done. 

Changing the match statement as suggested, didn't work, and I had to use map_err` to get the method to return the correct type of E in Result<T, E>. Let me know if there's an even sleeker way of achieving this.

eranrund replied:
Does `Ok(result?)` work?

samdealy replied:
Yes. Changed it to that. Thanks for suggestion, makes it sleek.

samdealy commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,546 @@
+        ))
+    }
+
+    fn activate_a_node(&self, activated_node_index: usize) -> Result<(), OverseerError> {
```
Done.

samdealy commented:
> Nice work. I have not looked at the tests, but I did have some (relatively minor) suggestions, and an open question about the use of stop_trigger. I think what should've happened is that arming starts the worker thread or errors if it is already started, and disarming stops it?

Ah I see your confusion. The way I've implemented it, the worker thread is started when OverseerService#start is called, which occurs when the "start_overseer_service" binary is executed. Arm is really only used after the service is disarmed- i.e. to restart the service after it's been stopped. In this sense, perhaps "arm" should be renamed to rearm? I can also modify it to emit an error if the worker thread is already running. 

That said, if you think the executable should not start the worker thread automatically let me know. I picked up this convention from other fog services, but perhaps it doesn't make sense here given that we are exposing an "arm" or "rearm" api.

eranrund commented:
> > Nice work. I have not looked at the tests, but I did have some (relatively minor) suggestions, and an open question about the use of stop_trigger. I think what should've happened is that arming starts the worker thread or errors if it is already started, and disarming stops it?
> 
> Ah I see your confusion. The way I've implemented it, the worker thread is started when OverseerService#start is called, which occurs when the "start_overseer_service" binary is executed. Arm is really only used after the service is disarmed- i.e. to restart the service after it's been stopped. In this sense, perhaps "arm" should be renamed to rearm? I can also modify it to emit an error if the worker thread is already running.
> 
> That said, if you think the executable should not start the worker thread automatically let me know. I picked up this convention from other fog services, but perhaps it doesn't make sense here given that we are exposing an "arm" or "rearm" api.

This still doesn't add up since I don't see where the thread is restarted... it appears its only started when the process starts.
Why do we need this if we are auto-starting it?

I also think that given what you said, the API names might just need to be "start" and "stop"?

eranrund commented:
And potentially include a status API endpoint that indicates if it is currently started...

eranrund commented on fog/overseer/server/src/service.rs:
```diff
@@ -66,15 +66,15 @@ where
-        log::trace!(self.logger, "Arming overseer worker");
-        self.stop_requested.store(false, Ordering::SeqCst);
+    pub fn enable(&self) -> Result<String, String> {
+        log::trace!(self.logger, "Enabling overseer worker");
```
Recommend `info!` so that there's a log of this happening.

samdealy replied:
Done.

eranrund commented on fog/overseer/server/src/service.rs:
```diff
@@ -66,15 +66,15 @@ where
-        log::trace!(self.logger, "Disarming overseer worker");
-        self.stop_requested.store(true, Ordering::SeqCst);
+    pub fn disable(&self) -> Result<String, String> {
+        log::trace!(self.logger, "Disabling overseer worker");
```
Recommend `info!` so that there's a log of this happening.

samdealy replied:
Done.

eranrund commented on fog/overseer/server/src/worker.rs:
```diff
@@ -32,8 +32,8 @@ pub struct OverseerWorker {
-    /// Stop request trigger, used to signal the thread to stop.
-    stop_requested: Arc<AtomicBool>,
+    /// When true, overseer performs its monitoring logic.
+    is_enabled: Arc<AtomicBool>,
```
Is there a need to store this here? If it comes from the outside (the caller of `OverseerWorker::new`) and is immediately passed to the thread, my thinking is no.

samdealy replied:
I moved it out. However, for the new AtomicBool, stop_requested, I keep it in this struct because the stop method needs  to set it to true.

eranrund commented on fog/overseer/server/src/worker.rs:
```diff
@@ -78,14 +78,14 @@ impl OverseerWorker {
     pub fn stop(&mut self) -> Result<(), ()> {
         if let Some(join_handle) = self.join_handle.take() {
-            self.stop_requested.store(true, Ordering::SeqCst);
+            self.is_enabled.store(false, Ordering::SeqCst);
```
Why would this cause the thread to stop?

samdealy replied:
Fixed in most recent commit.

eranrund commented on fog/overseer/server/src/worker.rs:
```diff
@@ -137,23 +137,28 @@ where
 
     fn run(self) {
-        while !self.stop_requested.load(Ordering::SeqCst) {
+        loop {
```
The thread is never going to terminate if the loop doesn't have an exit condition.

samdealy replied:
Fixed in most recent commit.

eranrund commented on fog/overseer/server/src/worker.rs:
```diff
@@ -137,23 +137,28 @@ where
             std::thread::sleep(Self::POLLING_FREQUENCY);
 
+            if !self.is_enabled.load(Ordering::SeqCst) {
+                log::info!(self.logger, "Overseer worker is currently disabled.");
```
This will be very spammy in the logs.

samdealy replied:
Fixed in most recent commit.

eranrund reviewed (changes requested):
It appears to me that the thread management is still buggy.
I am also worried that there are no tests that cover this, since if I am right something should've caught that - and if I'm wrong then I am misunderstanding something.

I also recommend adding a `/status` endpoint to check if the thing is enabled or disabled. This will likely be useful in the future.

samdealy commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,546 @@
+    /// cluster's ingress keys.
+    recovery_db: DB,
+
+    /// If this is true, the worker will not perform it's monitoring logic.
```
Fixed in most recent commit.

eranrund commented on fog/overseer/server/src/bin/main.rs:
```diff
@@ -0,0 +1,42 @@
+/// over HTTP.
+fn main() {
+    mc_common::setup_panic_handler();
+    // TODO: Enable sentry here.
```
Why not just stick that piece of code here? That's all that is needed to enable it.

samdealy replied:
I haven't set up sentry alerting yet and would like to do it all in one PR.

eranrund commented on fog/overseer/server/src/config.rs:
```diff
@@ -0,0 +1,58 @@
+    /// deployment.
+    ///
+    /// gRPC listening URIs for client requests.
+    #[structopt(long, use_delimiter = true)]
```
Still an open issue/question.

samdealy replied:
It seems easier to use a delimited set of words rather than typing in --peer multiple times, but can change it if you think that's better.

eranrund commented on fog/overseer/server/src/error.rs:
```diff
@@ -0,0 +1,47 @@
+pub enum OverseerError {
+    /// RecoveryDbError: {0}
+    RecoveryDb(Box<dyn RecoveryDbError>),
+    // TODO: Add more errors as appropriate
```
I don't think we need this comment.

samdealy replied:
Done.

eranrund commented on fog/overseer/server/src/service.rs:
```diff
@@ -0,0 +1,93 @@
+    overseer_worker: Option<OverseerWorker>,
+    recovery_db: DB,
+    is_enabled: Arc<AtomicBool>,
+    stop_requested: Arc<AtomicBool>,
```
No need to store this in this struct since it doesn't look like it's ever being accessed.

samdealy replied:
Nevermind, we can remove it.

eranrund commented on fog/overseer/server/src/service.rs:
```diff
@@ -0,0 +1,93 @@
+            self.recovery_db.clone(),
+            self.logger.clone(),
+            self.is_enabled.clone(),
+            self.stop_requested.clone(),
```
This can live entirely inside `OverseerWorker` and doesn't need to be exposed.

samdealy replied:
What can live inside it? (GitHub isn't showing me what you highlighted / are specifically referring to).

samdealy replied:
I see now- you're referring to self.stop_requested.clone(). Changed accordingly. Done.

eranrund commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,546 @@
+        }
+
+        // If there's an outstanding key, we need to report it as lost.
+        if let Some(inactive_outstanding_key) = inactive_outstanding_keys.get(0) {
```
Still an open question

garbageslam replied:
Why do we repeat this `if` test, can't we fold it into the earlier `if` ?

samdealy replied:
Good point. Done.

eranrund commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,544 @@
+                loop {
+                    log::info!(
+                        self.logger,
+                        "Trying to activate node at index {}",
```
I suggest logging the URI from the client as well, it will make it easier for anyone looking at the logs.

samdealy replied:
Done.

eranrund commented on fog/overseer/server/tests/inactive_outstanding_key_idle_does_not_have_key.rs:
```diff
@@ -0,0 +1,216 @@
+    overseer_service.start().unwrap();
+
+    // Set up the Rocket instance
+    let overseer_state = OverseerState { overseer_service };
+    // TODO: Consider testing the CLI here instead.
+    let rocket_config: rocket::Config =
+        rocket::Config::build(rocket::config::Environment::Development)
+            // TODO: Make these either passed from CLI or in a Rocket.toml.
+            .address("127.0.0.1")
+            .port(80)
+            .unwrap();
+    let rocket = server::initialize_rocket_server(rocket_config, overseer_state);
+    let client = Client::new(rocket).expect("valid rocket instance");
+    let _req = client.post("/arm").dispatch();
```
I'm surprised to see this code in the tests. Why can't it use the stuff inside `fog/overseer/server/src/server.rs`, or just operate on `overseer_service` directly?

samdealy replied:
I'd like to make it a proper end to end test. Why not exercise all of the code paths, rather than a layer beneath it? Without this, we can't be certain that the rocket logic is working correctly.

eranrund commented on fog/overseer/server/tests/inactive_outstanding_key_idle_does_not_have_key.rs:
```diff
@@ -0,0 +1,216 @@
+        rocket::Config::build(rocket::config::Environment::Development)
+            // TODO: Make these either passed from CLI or in a Rocket.toml.
+            .address("127.0.0.1")
+            .port(80)
```
This is fishy, ports lower than 1024 requires running as root.

samdealy replied:
Changed to 8080 and above.

eranrund commented on fog/overseer/server/tests/inactive_outstanding_key_idle_has_key.rs:
```diff
@@ -0,0 +1,208 @@
+        rocket::Config::build(rocket::config::Environment::Development)
+            // TODO: Make these either passed from CLI or in a Rocket.toml.
+            .address("127.0.0.1")
+            .port(80)
```
Using the same port in multiple tests can have undesirable effects such as one of the tests failing to bind the port, or the requests going to the wrong place.

Like in the previous comment, I am not sure there's a need to bother with rocket here at all.

samdealy replied:
See above.

eranrund commented on fog/overseer/server/tests/utils/mod.rs:
```diff
@@ -0,0 +1,172 @@
+// Add an arbitrary block to ledger and a timestamp for it
+// This method is used in integration tests, but cargo still complains that it's
+// not in use.
+#[allow(dead_code)]
```
Is this needed?

samdealy replied:
Yes, CI complains that it's not used (I don't think cargo checks that it's being used in the tests/ directory).

eranrund reviewed (changes requested):
Nice progress, thanks for dealing with the thread stuff!

samdealy commented on fog/overseer/server/src/service.rs:
```diff
@@ -0,0 +1,80 @@
+
+    pub fn stop(&mut self) -> Result<(), OverseerError> {
+        // This blocks on teardown of overseer_worker
+        self.overseer_worker = None;
```
Removed the struct field. Done.

samdealy commented on fog/overseer/server/src/service.rs:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    pub fn arm(&self) -> Result<String, String> {
+        log::trace!(self.logger, "Arming overseer worker");
```
This change was made in subsequent commit.

samdealy commented on fog/overseer/server/src/service.rs:
```diff
@@ -0,0 +1,80 @@
+
+    pub fn arm(&self) -> Result<String, String> {
+        log::trace!(self.logger, "Arming overseer worker");
+        self.stop_requested.store(true, Ordering::SeqCst);
```
Fixed in later commit.

samdealy commented on fog/overseer/server/src/service.rs:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    pub fn disarm(&self) -> Result<String, String> {
+        log::trace!(self.logger, "Disarming overseer worker");
```
Done.

samdealy commented on fog/overseer/server/src/service.rs:
```diff
@@ -0,0 +1,80 @@
+
+    pub fn disarm(&self) -> Result<String, String> {
+        log::trace!(self.logger, "Disarming overseer worker");
+        self.stop_requested.store(false, Ordering::SeqCst);
```
Fixed in later commit. Done.

eranrund commented on fog/overseer/server/src/service.rs:
```diff
@@ -56,7 +54,7 @@ where
             self.logger.clone(),
             self.is_enabled.clone(),
-            self.stop_requested.clone(),
+            Arc::new(AtomicBool::new(false)),
```
Why pass this as an argument? It's an internal implementation detail of `OverseerWorker`

samdealy replied:
Agreed. Done.

samdealy commented:
> I see, that makes sense. Does `Self::NUMBER_OF_TRIES as usize` doesn't work?

It does, and that's the way I had it written earlier. I find it a little bit cleaner to just write the `try_into` rather than the as usize during the subtraction. Happy to change it though, especially if it is more idiomatic that way.

samdealy commented:
@eranrund Github isn't letting me reply directly to the "use_delimiter" comment of yours, so will do so here

I'd prefer to use a delimiter because I think it's easier / less typing than putting` --ingest-cluster-uris` before each cluster uri. However, I think it's reasonable to use a delimiter if only because that's the convention in Fog. Happy to change it if you feel strongly about it.

eranrund commented:
> @eranrund Github isn't letting me reply directly to the "use_delimiter" comment of yours, so will do so here
> 
> I'd prefer to use a delimiter because I think it's easier / less typing than putting` --ingest-cluster-uris` before each cluster uri. However, I think it's reasonable to use a delimiter if only because that's the convention in Fog. Happy to change it if you feel strongly about it.

I dont feel strongly about it so whatever works for you.

samdealy commented:
> And potentially include a status API endpoint that indicates if it is currently started...

Done

eranrund commented:
@samdealy if you don't mind, I'm going to wait with my review until CI is green. Can you please ping me when its happy?

garbageslam commented on fog/overseer/server/src/service.rs:
```diff
@@ -0,0 +1,90 @@
+    pub fn disable(&self) -> Result<String, String> {
+        log::info!(self.logger, "Disabling overseer worker");
+        self.is_enabled.store(false, Ordering::SeqCst);
+        Ok(String::from("Fog Overseer was successfully disabled."))
```
It might be nice to make it like,

```
let old_value = self.is_enabled.swap(false, ...);
if old_value { "Fog Overseer was successfully disabled" } else { "Fog Overseer was already disabled" }
```

might not be super important but it might help if the user is having trouble using it

samdealy replied:
Done.

garbageslam commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,544 @@
+
+    /// Returns the latest round of ingest summaries for each
+    /// FogIngestGrpcClient that communicates with a node that is online.
+    fn retrieve_ingest_summary_node_mappings(&self) -> Vec<IngestSummaryNodeMapping> {
```
I think this should return the error, so that the caller can abort this pass if some of the nodes are not reachable / not responding?

garbageslam replied:
if any of the nodes are not reachable then the overseer should not do anything, otherwise if the network partitions bad things can happen

samdealy replied:
Yes good point. Originally, I was considering the case in which one of the nodes is continually unresponsive. In this scenario, there could be no active nodes and this unresponsive node could be an idle node. In this case, we'd want overseer to respond, but it wouldn't because of this one unresponsive node.

However, I agree with you that it's too dangerous to proceed with the logic if a node is unresponsive because it could be the active node and then come back online and there'd be two active nodes (a second one would be activated by overseer in the interim). 

I also think that because we're assuming that the unresponsive node will become responsive soon enough (i.e. Azure will bring the node back online in a relatively reasonable time frame), then the probability of this "infinite loop of inaction" is pretty low. 

I changed the method.

garbageslam commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,544 @@
+            }
+
+            let ingest_summary_node_mappings: Vec<IngestSummaryNodeMapping> =
+                self.retrieve_ingest_summary_node_mappings();
```
I would want to make it so that if some nodes are unreachable, we back off until the next polling cycle -- so this would check if any node gave an error, then we log it, and `continue` the loop

samdealy replied:
Done.

garbageslam commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,544 @@
+                    self.logger,
+                    "There was an issue activating a node for the outstanding key: {:?}",
+                    err
+                ),
```
i think perhaps we should not fall through here?

samdealy replied:
Agreed. Changed.

garbageslam commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,544 @@
+        }
+
+        if let Some(inactive_outstanding_key) = inactive_outstanding_keys.get(0) {
+            match self.try_to_activate_node_for_inactive_outstanding_key(
```
I think this `try_to_activate_node` function is a little too busy -- there are errors that can occur if contacting a node fails, and errors that can occur if finding the key fails it seems. This seems bad for clear error handling.

What I would suggest is that we try to move more logic from that function to this function, and possibly eliminate that function:

* You could iterate across the `ingest_summary_node_mappings`, using `iter().find()`, or a for loop, whatever you think is more clear
* Try to activate the first node that has the desired key. If that fails, return an error, and we'll re-evaluate everything and try again in the next pass
* If no node has the desired key, then we report a lost key, and try to activate the first node

In general any errors that we get when talking to ingest should not be just "log and fall through", they should cause us to abort, and return an error to the caller, who logs it. This will make it much easier to figure out in what circumstances the overseer may attempt to activate something. We want that circumstance to be as simple and as restrictive as possible. It should only do this when we are as certain as we can be that this is the right thing to do.

samdealy replied:
Agreed. Thank you for expressing this so clearly. Implemented it as you described.

garbageslam commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,544 @@
+                // time, we should just keep trying here.
+                //
+                // TODO: Add alerting if this is taking too long.
+                loop {
```
I think it would be better not to do this in a loop, we could just return the error to the caller?

Because the information we used to decide to activate this node could become stale -- what if the operator is also doing something in background for example. If it is in this loop for 20 minutes then the overseer might be attempting to do something that no longer makes sense

samdealy replied:
Agreed. As mentioned in an above comment, I moved this logic to the `perform_automatic_failover `method, and removed this loop. A failure to activate the node results in an error being propagated to the caller and no activation logic proceeds.

garbageslam reviewed (changes requested):
This looks pretty good but there are some things that should change:

* We want the overseer to be very conservative about when it decides to activate things. But there are some conditions that it doesn't handle well right now. If the network partitions and some of the ingests are not reachable, it logs those errors but it doesn't stop it from going forward and trying to compute the number of active nodes. That's dangerous because if you can't reach any of the servers then you don't actually know the number of active nodes.
* The `try_to_activate_node_for_inactive_outstanding_key` function is too complicated, it's returning basically two very different types of errors that have to be handled differently. One of them is that the key we need doesn't exist anymore, and one of them is that, asking the server that had it to activate has failed. These errors need to be handled differently by the caller but the current code isn't doing that. I think this code would be simpler if we essentially inlined this function into the caller and then handled these errors correctly.
* There are several places where we handle errors by logging them and falling through. I think this is wrong in all the places that I saw, and it makes it too complicated to figure out in what cases the overseer will activate something. All these places should be fixed by returning an error to the caller.
* I think the infinite loop where we try to activate something forever is also wrong. We could retry it 2 or 3 times, but after that, we should return an error to caller, who logs it, and re-evaluate the whole situation in the next cycle.

samdealy commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,546 @@
+        inactive_outstanding_key: &CompressedRistrettoPublic,
+        ingest_summary_node_mappings: &[IngestSummaryNodeMapping],
+    ) -> Result<(), OverseerError> {
+        log::debug!(
```
Done.

samdealy commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,546 @@
+            self.report_lost_ingress_key(*inactive_outstanding_key)?;
+        }
+
+        // Regardless of whether or not there's an outstanding key, we need
```
Good point. I changed the behavior to now return early if the key activation fails. So this code won't be reached in that case anymore.

samdealy commented:
> This looks pretty good but there are some things that should change:
> 
> * We want the overseer to be very conservative about when it decides to activate things. But there are some conditions that it doesn't handle well right now. If the network partitions and some of the ingests are not reachable, it logs those errors but it doesn't stop it from going forward and trying to compute the number of active nodes. That's dangerous because if you can't reach any of the servers then you don't actually know the number of active nodes.
> * The `try_to_activate_node_for_inactive_outstanding_key` function is too complicated, it's returning basically two very different types of errors that have to be handled differently. One of them is that the key we need doesn't exist anymore, and one of them is that, asking the server that had it to activate has failed. These errors need to be handled differently by the caller but the current code isn't doing that. I think this code would be simpler if we essentially inlined this function into the caller and then handled these errors correctly.
> * There are several places where we handle errors by logging them and falling through. I think this is wrong in all the places that I saw, and it makes it too complicated to figure out in what cases the overseer will activate something. All these places should be fixed by returning an error to the caller.
> * I think the infinite loop where we try to activate something forever is also wrong. We could retry it 2 or 3 times, but after that, we should return an error to caller, who logs it, and re-evaluate the whole situation in the next cycle.

Agreed with the points here. I implemented the changes listed in the above comments that address these points. I moved the `try_to_activate_node_for_inactive_oustanding_key` logic to the `perform_automatic_failover` method, and return early if a node that has the inactive outstanding key can't be activated successfully. 

I just double-checked, and I believe that this removes all of the situations in which we were logging and error and falling through (will check again on Monday). 

For the last point, I modified it to try and activate the node once, and if it doesn't work then it returns to the main thread loop. I can modify it to explicitly retry the activate call 2 to 3 times, but my sense is that it accomplishes basically the same the behavior as if we return to the loop (the loop will pause for 5 seconds according to it's sleep and then will retry activation). I'd be happy to modify it to explicilty retry 2 to 3 times though (which might make sense given that the rest of our node method calls explicitly do this too).

garbageslam commented on fog/overseer/server/src/lib.rs:
```diff
@@ -0,0 +1,10 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+#![feature(proc_macro_hygiene, decl_macro)]
```
do we really need nightly features for this?

can we avoid using unstable features if they aren't needed?

samdealy replied:
Yes, we need these to enable [rocket_codegen](https://docs.rs/rocket_codegen/latest/rocket_codegen/), which provides the macros for `#[post(...)] `and `#[get(...)]`. [Full Service](https://github.com/mobilecoinofficial/full-service/blob/main/full-service/src/lib.rs#L5) uses these declarations as well.

garbageslam replied:
ok

eranrund commented on fog/overseer/server/src/config.rs:
```diff
@@ -0,0 +1,51 @@
+            "--listen-port",
+            "8080",
+            "--ingest-cluster-uris",
+            "insecure-fog-ingest://0.0.0.0.3226/,insecure-fog-ingest://0.0.0.0.3227/",
```
This test is incomplete - notice how the port number is not delimited by `:` but the test did not actually catch that.

samdealy replied:
Added a test to address this.

eranrund commented on fog/overseer/server/src/service.rs:
```diff
@@ -0,0 +1,102 @@
+    pub fn disable(&self) -> Result<String, String> {
+        log::info!(self.logger, "Disabling overseer worker");
+        let was_enabled = self.is_enabled.swap(false, Ordering::SeqCst);
+        let response_message = if was_enabled {
```
Nit: a match statement like in `get_status` would be a bit shorter and easier to read (in my opinion)

samdealy replied:
Agreed. Done.

eranrund commented on fog/overseer/server/src/service.rs:
```diff
@@ -0,0 +1,102 @@
+    pub fn enable(&self) -> Result<String, String> {
+        log::info!(self.logger, "Enabling overseer worker");
+        let was_enabled = self.is_enabled.swap(true, Ordering::SeqCst);
+        let response_message = if was_enabled {
```
Nit: a match statement like in `get_status` would be a bit shorter and easier to read (in my opinion)

samdealy replied:
Agreed. Done.

eranrund commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,511 @@
+    where
+        OverseerError: From<DB::Error>,
+    {
+        let thread_is_enabled = is_enabled;
```
Nit: I dont think this is needed, can just use `is_enabled` inside the thread

eranrund commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,511 @@
+            _ => {
+                log::error!(self.logger, "Found multiple outstanding keys. This requires manual intervention. Disabling overseer.");
+                self.is_enabled.store(false, Ordering::SeqCst);
+                return Err(OverseerError::MultipleOutstandingKeys("Multiple outstanding keys found. This is unexpected and requires manual intervention. As such, we've disabled overseer. Take action and then enable overseer.".to_string()));
```
Nit: Maybe be worth including the keys?

eranrund replied:
Another nit is I've seen a mixture of `.to_string()`, `.to_owned()` and `String::from()`. They all do the same, I think `to_owned()` or `to_string()` are both used in the codebase, I would pick one and stick to it.

samdealy replied:
I included the keys and standardized on to_string.

eranrund commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,511 @@
+                            );
+                            return Ok(());
+                        }
+                        Err(_) => {
+                            let error_message =
+                                format!("Tried activating node {}, but it failed.", node.get_uri());
```
```suggestion
                        Err(err) => {
                            let error_message =
                                format!("Tried activating node {}, but it failed: {}", node.get_uri(), err);
```

(not tested)

samdealy replied:
Done.

eranrund commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,511 @@
+            }
+        }
+
+        if let Some(inactive_outstanding_key) = inactive_outstanding_keys.get(0) {
```
Nit: I personally am finding this trick a bit confusing. I would change the match statement above to do `        let activated_node_index = self.set_new_key_on_a_node()?; self.activate_a_node(activated_node_index)?;` on the zero case, and then have the rest of the logic under the `1` case. I understand it will result in repeating the two lines mentioned here but I think overall it will be easier to follow.

samdealy replied:
I agree that this is confusing. I implemented as you mentioned and pulled out the one outstanding key logic into its own method. Thank you for the suggestion!

eranrund replied:
Nice, that looks great. Thank you!

eranrund commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,511 @@
+                    let number_of_remaining_tries = Self::NUMBER_OF_TRIES - current_try;
+                    let error_message = match number_of_remaining_tries {
+                        0 => format!("Did not succeed in reporting lost ingress key {} within {} tries.", inactive_outstanding_key, Self::NUMBER_OF_TRIES),
+                        _ => format!("The following key  was not successfully reported as lost: {}. Will try {} more times", inactive_outstanding_key, number_of_remaining_tries),
```
```suggestion
                        _ => format!("The following key was not successfully reported as lost: {}. Will try {} more times", inactive_outstanding_key, number_of_remaining_tries),
```

eranrund replied:
Might be worth to also include the error in the formatting.

samdealy replied:
Done. Added underlying error message to retry error messages.

eranrund reviewed (approved):
Bunch of small nits but nothing that I feel too strongly about. Thank you!a

garbageslam commented on fog/overseer/server/src/lib.rs:
```diff
@@ -0,0 +1,10 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+#![feature(proc_macro_hygiene, decl_macro)]
+
```
nit: This crate should have `#[deny(missing_docs)]`, ideally all our crates would have that

samdealy replied:
Done.

garbageslam commented on fog/overseer/server/src/error.rs:
```diff
@@ -0,0 +1,52 @@
+    MultipleOutstandingKeys(String),
+
+    /// Unknown error associated with retries. Should not happen: {0}
+    GenericRetryError(String),
```
I think we probably don't need this? The internal retry errors actually never happen:

https://github.com/jimmycuadra/retry/issues/38

samdealy replied:
Done. Thanks for the context!

garbageslam commented on fog/overseer/server/src/error.rs:
```diff
@@ -0,0 +1,52 @@
+                total_delay: _,
+                tries: _,
+            } => error,
+            RetryError::Internal(s) => Self::GenericRetryError(s),
```
i think its okay to just panic in this case

samdealy replied:
Done.

garbageslam commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,543 @@
+                    let error_message = format!(
+                        "Unable to retrieve ingest summary for node: {}",
+                        ingest_client.get_uri()
+                    );
```
it would be better to capture the actual error and format it into this string, so like:

```
                Err(err) => {
                    let error_message = format!(
                        "Unable to retrieve ingest summary for node ({}): {}",
                        ingest_client.get_uri(),
                        err
                    );
```

or something like this

samdealy replied:
Done.

garbageslam reviewed (approved):
This all looks good to me, thank you! I had only minor comments

jcape commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,544 @@
+            let active_node_count = active_ingest_summary_node_mappings.len();
+            match active_node_count {
+                0 => {
+                    log::error!(
```
Recommend `warn!` here, at best. Situation is abnormal, but no human intervention is required (yet).

samdealy replied:
SGTM, done.

jcape commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,544 @@
+
+    /// Try a request to Fog Ingest node this many times if you encounter an
+    /// error.
+    const NUMBER_OF_TRIES: u64 = 3;
```
Is there any reason for this to be a `u64` instead of a `usize`? It would let you remove all the `try_into().unwrap()` calls later...

samdealy replied:
Yes, the reason is that the `current_try` parameter is `u64` (kinda weird to me that the retry library defines the types this way). 

I was talking with eran about this very concern and weakly prefer this to writing `as` when I do the subtraction to see how many calls remain (see line 523). that said, will change it if you'd like.

jcape replied:
I'm ambivalent, but there's definitely a strong preference against `unwrap()` dating back to the early days of our codebase ;-)

I suppose the main thing is: `usize as u64` is infallible on all platforms today, whereas (obviously) `try_into().unwrap()` is not.

samdealy replied:
Thanks for the context. Didn't realize that last point. Implemented it as you suggested. Done.

jcape reviewed (approved):
LGTM generally, though in the interests of time, I didn't review the integration tests, and most of my comments are purely stylistic.

One suggestion I would have for this and future (large) PRs is adding a writeup of the design as a section in `README.md`---like, the code seems reasonable, but I wasn't involved in the design beyond a preliminary sketch many months ago, so there's always going to be a "fit for purpose" gap without that getting written down.

samdealy commented on fog/ingest/server/src/server.rs:
```diff
@@ -359,6 +359,19 @@ where
     }
 
+    /// Ask if the server is idle.
+    /// This is a convenience wrapper used in tests.
```
That sounds good to me- didn't realize you could add this to a method. I went ahead and added it to other Fog Ingest sever methods that are only meant for tests.

garbageslam replied:
I think its okay, we have similar conventions in fog ingest to expose some state queries like this so that you don't need to use RPC for absolutely everything.

If we make it `cfg(feature = "under_test")` then you will have to rebuild the entire server to run the test which costs developer time. Meanwhile in the release mode binary this will get stripped out by the linker since it's not called in release mode.

samdealy replied:
Good call. Can confirm that the overseer integration tests fail if you add this attribute to the methods (compiler says it can't find the methods). I'm going to take them out.

samdealy commented on fog/ingest/server/src/server.rs:
```diff
@@ -359,6 +359,19 @@ where
+
+    /// Set new keys.
+    /// This is used in tests when it would be simpler than making an RPC
+    /// client.
```
Done.

samdealy commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,536 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+use crate::error::OverseerError;
+
```
Done.

samdealy commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,536 @@
+    logger: Logger,
+}
+
+/// This associates an IngestSummary with an IngestClient. This makes it easy
+/// to query a given node based on its IngestSummary.
+struct IngestSummaryNodeMapping {
+    node_index: usize,
+
+    ingest_summary: IngestSummary,
+}
```
yeah that would be more compact, and earlier I wasn't sure if I should be adding these derives to IngestSummary.

That said, I weakly prefer keeping it as is because it allows us to name what the `usize` parameter is. Without this named `node_index` field, I could foresee readers having a hard time understanding what the usize means. That said, I'd be happy to change it. LMK.

jcape replied:
I don't have a strong preference, how often are queries (map) vs. iterations (vec) is really the determining factor...

samdealy replied:
Good point. We're doing more iterations than queries, so will leave it as such.

samdealy commented on fog/overseer/server/src/lib.rs:
```diff
@@ -0,0 +1,10 @@
+
+#![feature(proc_macro_hygiene, decl_macro)]
+
+pub mod config;
+pub mod server;
+pub mod service;
```
I agree. However, I think the `pub` modifier (which i think == export?) is needed here because the integration tests in the `test/` need to access these modules. If I remove `pub` for any of these modules it doesn't compile.

samdealy commented on fog/overseer/server/src/worker.rs:
```diff
@@ -0,0 +1,544 @@
+                                    .get_ingress_pubkey()
+                            })
+                            .collect();
+                    log::warn!(
```
Yes good point. I updated this downstream in my forthcoming Sentry PR. Makes sense to just do it here. Done.

garbageslam commented:
FWIW I also like james' request to add a README.md

samdealy commented:
> FWIW I also like james' request to add a README.md that gives some kind of overview of the project and its goals

I just added a README.

samdealy commented:
> LGTM generally, though in the interests of time, I didn't review the integration tests, and most of my comments are purely stylistic.
> 
> One suggestion I would have for this and future (large) PRs is adding a writeup of the design as a section in `README.md`---like, the code seems reasonable, but I wasn't involved in the design beyond a preliminary sketch many months ago, so there's always going to be a "fit for purpose" gap without that getting written down.

Just added a README.
